Add initial versions for charts leading into plotting functionality

t-kramer opened the pull request:
PR introduces preliminary versions for a first set of charts that can be used along and/or integrated with pythermalcomfort. Before we improve the charts, clean the code and add more, we should discuss how we want to integrate such plots in general. See more on this below in the Motivation bit.

## Changes
- add `charts.py`: contains initial code for an Adaptive chart (+ alternative version with timeseries summary), PMV chart, and a Heatmap (again with a summary option)
- add `theme.py`: defines a Plotly template, enhancing consistency and streamlining plot functions - very basic for now, but can be leveraged much more
- add `chart_development.ipynb`: loads and preprocesses datasets used as input for charts (to be removed later)
- add 'test_data` folder: stores EPW files for testing (to be removed later)

### Added dependencies
- https://github.com/building-energy/epw: to process EPW files

## Screenshots
Images of charts created using Plotly: Most of them are based of code that I've written earlier and some functions that we used for Clima and the new Comfort Tool version. **All plots are interactive**, so they provide additional information when hovering over them, e.g. exact x, y, z data behind plotted data points.

![adaptive](https://github.com/user-attachments/assets/9947ca8c-60ac-485c-a133-b6cba87f0828)

![adaptive_with_summary](https://github.com/user-attachments/assets/6f381ae7-3f05-4cc7-a96d-39a1cbe8ec35)

![pmv](https://github.com/user-attachments/assets/485702be-9ace-4fc4-ac25-8c1c547aa326)

![heatmap](https://github.com/user-attachments/assets/d3ba7368-ef85-4795-9ef3-ba7f01b2bc37)

![heatmap_with_summary](https://github.com/user-attachments/assets/6bf70537-faa0-4f9a-bcfe-72d764771e73)

## Motivation / Ideas behind the charts
- big strength of pythermalcomfort over the CBE Comfort Tool: people can use it with multiple samples of input data instead of just one configuration
- makes the package very interesting for energy modelers and other AEC people, e.g. analyzing time series data
- plots should focus on these applications

- in pythermalcomfort, we have mainly three types of functions: PMV-based indexes, Adaptive indexes and others who either generate a temperature-based output or 'comfort' category (e.g. Heat Index)
- for a start, we could create one plot for each of these types: summarize time-series data, e.g. hourly simulation or measurement results and longitudinal field study data
- heatmap is very flexible: allows to plot the data (continuous or categorical) for most of the indices we offer
- added version with a little subplot on the right: summary is important for AEC people, especially when comparing different datasets - visually, this is often not clear enough and needs to be backed by some numbers, e.g. in per cent or hour counts
- can add something similar to the PMV plot, too

## (Potential) Next Steps

- [ ] visually further enhance charts
- [ ] develop alternative for psychrometric chart / PMV chart
- [ ] create one more chart as an alternative to the heatmap, maybe something that doesn't necessarily focus on mapping timeseries data
- [ ] clean code and improve modularity and versatility

## Up for discussion - How do we want to integrate this?
Initially, I was leaning towards integrating the plotting functionality as much as possible with the main codebase, for example using dot notation. However, I changed by mind a bit. I think especially when it comes to plotting, we should grant lots of flexibility. Pythermalcomfort is great, because it simplifies peoples workflows significantly (since they don't have to write all the included functions themselves), but also offer lots and lots of flexibility on how to use these functions.

So for everything that goes beyond the main codebase and it's functions, e.g. the plots, we could just provide a few examples of how to use pythermalcomfort on typical datasets in Pandas dataframe format (the examples don't do that yet, although I think it's the main use case) and provide users with a few insightful and nicely formatted plots, wrapped in functions they can manipulate easily if they want. Internally, we can keep curating these plots and update them on a rolling basis, for example when we create new plots using pythermalcomfort for data processing for papers or our tools. Considering the variety of applications of pythermalcomfort, writing some highly specific plots will limit their application in practice.

@FedericoTartarini @stefanoschiavon 

Keen to hear your feedback and thoughts. 🙋🏼‍♂️ 

\T

t-kramer committed:
feat: create first charts

FedericoTartarini commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+    return fig
+
+
+def pmv_chart(df, pmv_constants):
```
add doc string to explain what the inputs are. You can use the doc strings of the functions in models

t-kramer replied:
Done. Docstrings are preliminary and will need updates after functions are finalized.

FedericoTartarini replied:
let's then add a 
` TODO: change the docstring ....` after the docstring so pycharm highlights things that need to be completed

FedericoTartarini commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+)
+
+
+def adaptive_chart(df):
```
we should discuss whether this df should already contains the results of the adaptive model. so we can simplify significantly the code below.

FedericoTartarini commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+        RH, Tair = zip(*zone)
+        return Tair, [convert_rh_to_hr(T, rh) for T, rh in zip(Tair, RH)]
+
+    target_pmvs = {"Neutral": 0, "Upper": 0.5, "Lower": -0.5}
```
this shoudl be an input and maybe three separate variables.

FedericoTartarini commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+    return fig
+
+
+def heatmap(df, var, global_local, si_ip):
```
we need to discuss about the si_ip functionality, if we want to include it, then we should include it in all the charts. Perhaps it is necessary to include it to help people in the US.

t-kramer replied:
I think we have to implement it (but I haven't yet). Should be easy to do if we just convert the units for the output and leave all function in SI units.

FedericoTartarini commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+    return fig
+
+
+def heatmap_with_summary(df, var, global_local, si_ip):
```
this should not be a separate function from the previous but they should use the same codebase to avoid duplications.

t-kramer replied:
Yes, agreed and implemented.

FedericoTartarini commented on pythermalcomfort/theme.py:
we should discuss how we enable people to change this file so they can adapt the figures to their needs.

t-kramer replied:
Yes, I see this as the main way people can make customizations, because it does not blow up the functions definitions with lots of variables. Can also add other settings here, e.g. figsize.

FedericoTartarini commented:
@t-kramer great work and thank you for contributing. You have done a great work and I am very happy to see the charts implemented in pythermalcomfort. 

I have a few comments below and inline in the code in the pull request.

# code

1. I would create a subdirectory called `charts` which will contain the code for all the charts. Please add all the files here `pythermalcomfort/charts`
2. I would create a Python file for each chart, using the same approach used for the models, where each model has its own file.
3. `chart_development.ipynb` can be moved to the `examples` directory. There we can showcase a few charts. However, we need to make sure that we provide examples so all the charts can be generated without the need of using a Jupiter notebook. I really do not like jupiter files since they do not even render here in GitHub and are impossible to review. 
4. `test_data` can also be move to the examples folder
5. epw could be a dev dependency.
6. for now do not implement more chart since we want to first consolidate how results are calculated and then how charts are called.

# charts

the chart look great and very modern.

1. I am not sure what the round things in chart 2 mean (applicability, acceptability)
2. we need to make sure that people can pass the chart size as input so the font can be scaled by python.
3. PMV and adaptive charts - it would be nice to add a count of the datapoints within and outside the comfort region. Similar to chart 5.
4. Chart 4 and 5 are the same. By default we should show chart 5 and then people can hide the bar chart if they want.

# motivation

> heatmap is very flexible: allows to plot the data (continuous or categorical) for most of the indices we offer

agree also with the following comment, the bar chart summary is essential, the heatmap alone is beautiful but useless. We should also consider different grouping, month, time, etc.

> for a start, we could create one plot for each of these types: summarize time-series data, e.g. hourly simulation or measurement results and longitudinal field study data

# next step

1. add the T, RH chart.
2. discuss how the code is called and how people can use it.

> I think especially when it comes to plotting, we should grant lots of flexibility ... we could just provide a few examples of how to use pythermalcomfort on typical datasets in Pandas dataframe format (the examples don't do that yet, although I think it's the main use case) and provide users with a few insightful and nicely formatted plots, wrapped in functions they can manipulate easily if they want.

agree, we should probably do not include it with dot notation.

t-kramer commented on Pipfile:
```diff
@@ -25,4 +27,4 @@ bump-my-version = "*"
 [requires]
 python_version = "3.12"
-python_full_version = "3.12.4"
+python_full_version = "3.12.5"
```
I had issues with pipenv, this is why I changed it. Reverted it back to original line.

t-kramer commented on Pipfile:
```diff
@@ -9,6 +9,8 @@ numba = "*"
 numpy = "*"
 setuptools = "*"
 tox = "*"
+black = "*"
```
Done.

t-kramer commented on pythermalcomfort/charts.py:
Done.

t-kramer commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+
+def adaptive_chart(df):
+
+    OUTDOOR_TEMP_RANGE = (10, 33.5)
```
Yes to PEP! I introduced this check because I realised that the applicability check in the adaptive function is only implemented for `t_cmf` and not other outputs like `acceptability_80`. Let's check again and add it if not implemented. Removed the check for now.

t-kramer replied:
Regarding the first comment: We should discuss in detail how much direct customisability we want to grant folks. I'd suggest to provide indirect access to these parameters, e.g. via `theme.py` instead of getting all those settings from individual arguments in the function. We can talk about it.

FedericoTartarini replied:
Agree

t-kramer commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+    sum_80 = applicable_df["adaptive_acceptability_80%"].sum()
+    sum_90 = applicable_df["adaptive_acceptability_90%"].sum()
+
+    T_outdoor = np.linspace(OUTDOOR_TEMP_RANGE[0], OUTDOOR_TEMP_RANGE[1], 100)
```
Done, sorry for inconsistency.

t-kramer commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+
+    fig.add_trace(
+        go.Scatter(
+            x=np.concatenate([T_outdoor, T_outdoor[::-1]]),
```
I don't see a problem here, since I am not slicing a section of the data or pick an individual item of the list. Just reversing the list like this should be robust enough.

FedericoTartarini replied:
a okay, I did not understand that you are reversing the list. we can keep the code as it is but maybe add a comment explaining what we are trying to achieve.

t-kramer commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+            x=np.concatenate([T_outdoor, T_outdoor[::-1]]),
+            y=np.concatenate([T_90_upper, T_90_lower[::-1]]),
+            fill="toself",
+            fillcolor="rgba(0, 128, 0, 0.2)",
```
Similar to above, maybe we can provide that access via the dict in `theme.py` to keep the list of arguments in the functions small. Happy to implement a better solution if you have an idea!

t-kramer commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+    return fig
+
+
+def pmv_chart(df, pmv_constants):
```
Done.

t-kramer commented on pythermalcomfort/charts.py:
```diff
@@ -0,0 +1,624 @@
+
+def pmv_chart(df, pmv_constants):
+
+    T_dry = np.linspace(10, 36, 1000)
```
Reduced to 500.

t-kramer commented on chart_development.ipynb:
Done.

t-kramer committed:
fix: address comments from PR #161 review

This commit addresses the comments from PR #161 reviewed by
@FedericoTartarini. The following changes were made:

- improved and added summary for all charts
- created new modules for charts in pythermalcomfort/charts
- replaced chart_development.ipynb with example scripts and move these
to pythermalcomfort/examples
- move theme.py to pythermalcomfort/charts
- additional minor fixes

t-kramer commented:
@FedericoTartarini I finished most of my work yesterday, but didn't have the time to update the PR, so here it comes.

# Adjustments
I implemented most of your requested changes.
- improved and added summary for all charts
- created new modules for charts in pythermalcomfort/charts
- replaced `chart_development.ipynb` with example scripts and moved these
to pythermalcomfort/examples
- moved `theme.py` to pythermalcomfort/charts
- additional minor fixes

More information below and in the responses to your inline comments.

# Charts
> 1. I am not sure what the round things in chart 2 mean (applicability, acceptability)

This was another way of summarizing the data based on [this](https://escholarship.org/uc/item/7b4909sf) paper and Stefano's suggestion. The outer circles represent 100% and the inner circles the actually achieved number. Anyway, I removed this style and added the horizontal bar chart summary for all (see screenshot).

> 2. we need to make sure that people can pass the chart size as input so the font can be scaled by python.
>3. PMV and adaptive charts - it would be nice to add a count of the datapoints within and outside the comfort region. Similar to chart 5.
>4. Chart 4 and 5 are the same. By default we should show chart 5 and then people can hide the bar chart if they want.

Agree on all of these points, changes have been implemented accordingly.

Here are the updated charts (all including the summary, which can be hidden):

<img width="855" alt="Screenshot 2025-01-11 at 6 47 21 PM" src="https://github.com/user-attachments/assets/34e5842c-6f9b-49fa-8272-57c439755805" />

<img width="894" alt="Screenshot 2025-01-11 at 6 47 30 PM" src="https://github.com/user-attachments/assets/bbdcee21-0990-4b35-b4f9-2ef07be2a334" />

<img width="908" alt="Screenshot 2025-01-11 at 6 47 41 PM" src="https://github.com/user-attachments/assets/f28ada9a-a20e-40aa-953e-2548d376f230" />

<img width="897" alt="Screenshot 2025-01-11 at 6 47 49 PM" src="https://github.com/user-attachments/assets/8a6b7973-bb0c-4770-9bb9-90ac36f4d129" />

I know, the bar chart for the UTCI seems a bit crowded. Need to find a way how to handle this in a flexible way. Also, the missing data points (white pixels) are those that are out of range and therefore NaN.

Further, I do not agree completely with this comment:
> the bar chart summary is essential, the heatmap alone is beautiful but useless

While the heatmap alone can't give you a detailed summary of the whole year, it is very helpful in pointing out specific design problems when plotting indoor environmental data, e.g. too much solar radiation causing overheating at very specific hours of the year. So for some cases, people might want to use the heatmap without the summary.

# Next Steps

> 1. add the T, RH chart.
> 2. discuss how the code is called and how people can use it.

Let's do it! 🚀 

Thanks again for providing the detailed review, @FedericoTartarini! ❤️ 
Let me know if you have any further changes to implement before merging.

FedericoTartarini commented:
Great work. I left a few more comments in line with the code. Below a response to some of your comments and a few observations about the chart. I love how the charts look.

1. adaptive chart - it says that 0.0% of the data are out of range however this is not true. We should probably say < 0.1%
2. In the adaptive chart, should we not be showing the bars in the following order? Acceptable, acceptable for 80%, discomfort and out of range?
3. I've noticed that the chart size is still not an input of the function.
4. In the PMV chart the X axis and the Y axis have a solid line, but in the adaptive this is not the case.
5. In the PMV chart should be considering allowing to colour the dot as a function of the PMV value?
6. In the UTCI chart should not be the green in the center of the color bar?
7. Heatmap alone - OK let's then allowed the user to choose whether they want to see or not the summary.

Thank you so much for working on this. I really appreciate your help.

I have noticed that I can no longer see all my all comments, however, only a few of them have not been addressed yet.

FedericoTartarini commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+        var_metadata[si_ip]["unit"],
+        var_metadata[si_ip]["range"],
+    )
+    cat_color, cat_order = (
```
I believe these should be functions within the class and we should not be using strings. Let's discuss this during the meeting.

FedericoTartarini commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+            y=df["top"],
+            mode="markers",
+            name="Observations",
+            marker=dict(color="#ea536e", size=5, symbol="circle"),
```
Colors should also be defined in the configuration file

t-kramer replied:
Yes, I agree. I defined them locally when tweaking the functions. Enums?

FedericoTartarini replied:
Let's discuss this tomorrow, but they should be editable

FedericoTartarini commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+        category_counts = df["categorical"].value_counts()
+        category_percentages = df["categorical"].value_counts(normalize=True) * 100
+
+        categories = cat_order
```
I am not sure either why we are having a copy of a variable

FedericoTartarini commented on pythermalcomfort/charts/heatmap.py:
```diff
@@ -0,0 +1,215 @@
+    ----------
+    df : DataFrame
+        A DataFrame containing the input data. Must have the following columns:
+        'DateTime' (Date and time), 'Hour' (Hour of the day), 'Month Name' (Month name), 'Day' (Day of the month),
```
Please add the each column name on a different row if possible

FedericoTartarini replied:
Do we really need all these inputs? Can't we extract the our month name and the day from the date time?

FedericoTartarini commented on pythermalcomfort/charts/heatmap.py:
```diff
@@ -0,0 +1,215 @@
+    var_color = index_mapping_dictionary[var]["colors"]
+
+    if global_local == "global":
+        range_z = var_range
```
I don't think we should be using this code, but instead we should allow the user to pass the min and max value. If they don't, then we can use the values from the data set.

FedericoTartarini commented on pythermalcomfort/charts/heatmap.py:
```diff
@@ -0,0 +1,215 @@
+        data_min = 5 * floor(df[var].min() / 5)
+        range_z = [data_min, data_max]
+
+    if not show_summary:
```
Can't we flip the code so we avoid using `not show_summary`? A negation is always a bit more difficult to understand. What I'm proposing is to have the code if show summary is true first.

FedericoTartarini commented on pythermalcomfort/charts/heatmap.py:
```diff
@@ -0,0 +1,215 @@
+                zmin=range_z[0],
+                zmax=range_z[1],
+                customdata=np.stack(
+                    (df["Month Name"], df["Day"], df["Hour"], df["Hour"] + 1), axis=-1
```
We should never be using strings inside the code

FedericoTartarini commented on pythermalcomfort/charts/heatmap.py:
```diff
@@ -0,0 +1,215 @@
+        fig.update_layout(margin=tight_margins, width=800, height=500)
+
+    else:
+        cat_color = index_mapping_dictionary[var]["colors_categories"]
```
Same as the previous comment about the string. As previously mentioned, the categories should be extracted using a function within data class.

FedericoTartarini commented on pythermalcomfort/charts/pmv_chart.py:
```diff
@@ -0,0 +1,327 @@
+    ----------
+    df : DataFrame
+        A DataFrame containing the input data. Must have the following columns:
+        'ta' (Dry bulb temperature [°C]), 'rh' (Relative humidity [%]).
```
Throughout pythermalcomfort we are calling the temperature `tdb` not `ta` so we should be consistent

t-kramer replied:
I agree and adjusted that.

FedericoTartarini commented on pythermalcomfort/charts/pmv_chart.py:
```diff
@@ -0,0 +1,327 @@
+        A DataFrame containing the input data. Must have the following columns:
+        'ta' (Dry bulb temperature [°C]), 'rh' (Relative humidity [%]).
+
+    pmv_constants : dict
```
I am not a huge fan of dictionaries and dictionaries as inputs

FedericoTartarini commented on pythermalcomfort/charts/pmv_chart.py:
```diff
@@ -0,0 +1,327 @@
+        'tr' (Mean radiant temperature [°C]), 'v' (Air speed [m/s]), 'met' (Metabolic rate [met]), 'clo' (Clothing
+        insulation [clo]).
+
+    show_summary : bool, optional
```
I believe that the show summary should be through by default

t-kramer replied:
Implemented for all relevant functions.

FedericoTartarini commented on pythermalcomfort/charts/pmv_chart.py:
```diff
@@ -0,0 +1,327 @@
+
+    # helper functions start
+
+    def pmv_eq(T, RH):
```
In pythons to the variable name should not be capitalized

FedericoTartarini replied:
Do we really need this function to be defined here? can't we just have the PMV function in line 84?

FedericoTartarini commented on pythermalcomfort/charts/pmv_chart.py:
```diff
@@ -0,0 +1,327 @@
+        [[convert_rh_to_hr(T, RH) for T in t_dry] for RH in rh_levels]
+    )
+
+    target_pmvs = {"Neutral": 0, "Upper": 0.5, "Lower": -0.5}
```
This should be an input

FedericoTartarini commented on pythermalcomfort/charts/theme.py:
```diff
@@ -0,0 +1,140 @@
+
+tight_margins = dict(l=35, r=35, t=45, b=20)
+
+index_mapping_dictionary = {
```
We should not be using dictionary about the data classes instead

FedericoTartarini reviewed (commented):
I reviewed most of the code, and I left a few comments in line. Thank you so much for working on this.

One thing that a little bit worries me is that the some of my old comments are lost, but they were not addressed in this commit. Do you know what may have happened? We can discuss the code the during the upcoming meeting.

FedericoTartarini commented:
@t-kramer thank you for dedicating sometime to work on this before the meeting. Don't worry about the comments. I think they are all there. Perhaps I forgot to add a few last time. You did an amazing job replying to my comments, so don't worry. Sorry, sometimes for being annoying and asking you to change some stuff, but I just want to make sure that the code is easy to read and consistent.

I generally have the comments from GitHub, simply because I am lazy, and I've never learned how to do it inside by PyCharm. I am actually feeling embarrassed about this as I am typing the message. This is my new year resolution :)

FedericoTartarini commented:
@t-kramer I also forgot to mention that it would be great if you could extract the thermal comfort function from the plotting function. For example, considering that we have a different PMV model, we don't want to have a chart function for each specific formulation. It is better that the user passes in a column with the PMV results and then after the code does the plotting.

t-kramer committed:
build: merge updated upstream into plotting-feature

Merge was necessary to update the plotting-feature
branch with the latest changes from the upstream
development branch

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+    ----------
+    df : DataFrame
+        A DataFrame containing the input data. Must have the following columns:
+        't_out' (Outdoor temperature [°C]), 'top' (Operative temperature [°C]),
```
Done.

t-kramer commented on examples/plot_pmv.py:
```diff
@@ -0,0 +1,27 @@
+    low_memory=False,
+).reset_index(drop=True)
+
+pmv_subset = ashrae_measurements.loc[:, ["ta", "rh"]].dropna()
```
Done.

t-kramer commented on examples/plot_pmv.py:
```diff
@@ -0,0 +1,27 @@
+
+df = pmv_subset.sample(500, random_state=40)
+
+pmv_inputs = {
```
Instead of a this dict, we could also use individual arguments for each of these settings? What do you think? E.g. met_constant, clo_constant etc. We need to be careful not to confuse the users, since these inputs "only" define the comfort polygon.

t-kramer replied:
Regarding your second point: I know, and this is not the plan here. While the initial data from the ASHRAE database contains clo and met, I am only using a subset incl. tdb and rh here to test the function. What we could do here to avoid confusion is to provide a different example file, e.g. a typical simluation output file that contains zone tdb and rh only. I think using this function with simulation data will be the most relevant use.

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+)
+
+
+def adaptive_chart(df, show_summary=False, si_ip="si"):
```
I think this is not necessary. As long as we are consistent (e.g. use 'tdb' instead of 'ta'), we can expect people to be consistent with their column names, too.

FedericoTartarini replied:
perfect we need to be then very consistent and explain that in the docstrings

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+)
+
+
+def adaptive_chart(df, show_summary=False, si_ip="si"):
```
I am torn about this. While I think we should provide some flexibility (e.g. use of different color palettes), I think we can't let the users control everything. Especially the size of the plot is a very sensible parameter, which can mess up the chart pretty quickly. We should, however, pick a good size, e.g. optimize width and font size for scientific papers and/or presentation slides. Let's discuss this.

FedericoTartarini replied:
I completely understand your point but I belive people should be able to adjust it based on their needs. Not all scientific papers have the same column width or text width. I almost prefer that the chart is not going to look nice if the user selects very strange figure sizes, but at least they can see the results and they can customize it. Forcing just two or three figure sizes I don't think its ideal.

t-kramer replied:
We can implement that by passing an argument to the instance of the Base() class when building the plot.

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+        'adaptive_acceptability_80%' (Acceptability for 80% of the occupants),
+        'adaptive_acceptability_90%' (Acceptability for 90% of the occupants).
+
+    show_summary : bool, optional
```
Done.

t-kramer commented on examples/plot_adaptive.py:
```diff
@@ -0,0 +1,41 @@
+
+adaptive_subset = ashrae_measurements[["ta", "tr", "t_out", "vel"]].dropna()
+
+adaptive_subset["top"] = operative_tmp(
```
No, with this dataset we can't, since the outdoor temperature data is discontinuous. Similar to `plot_pmv.py`, we should probably use another example dataset here that has continuous outdoor temperature data that we can use to calculate the running mean.

FedericoTartarini replied:
Yes, let's use a better data set for the example

t-kramer commented on examples/plot_adaptive.py:
```diff
@@ -0,0 +1,41 @@
+)
+
+adaptive_results = adaptive_ashrae(
+    tdb=adaptive_subset["ta"].to_numpy(),
```
Yes, I think this is the way to go to avoid this error:
> "TypeError: tdb must be one of the following types: (<class 'float'>, <class 'int'>, <class 'numpy.ndarray'>, <class 'list'>)". So yes, let's fix the code to be able to pass pandas columns directy. Can we just allow pd.Series and convert it to numpy.ndarray?

t-kramer replied:
Issue has been resolved and code adjusted.

t-kramer commented on examples/plot_adaptive.py:
```diff
@@ -0,0 +1,41 @@
+    low_memory=False,
+).reset_index(drop=True)
+
+adaptive_subset = ashrae_measurements[["ta", "tr", "t_out", "vel"]].dropna()
```
Sure. We do you want to add this, here or within the function?

FedericoTartarini replied:
here since this is only relevant to the example, is not it? In the function documentation we need to state if we want a dataframe as input with or without nan

t-kramer commented on examples/plot_adaptive.py:
```diff
@@ -0,0 +1,41 @@
+import sys
+import os
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
```
Yes, the first four lines can go eventually. I'm just having some problem with the system path, maybe we can resolve it together. You used something similar in `docs/conf.py`, maybe we can fix this issue globally.

FedericoTartarini replied:
yes, let's try to fix it globally since it does not look very nice here at the beginning of the example

t-kramer commented on examples/plot_adaptive.py:
```diff
@@ -0,0 +1,41 @@
+
+from pythermalcomfort.utilities import operative_tmp
+from pythermalcomfort.models import adaptive_ashrae
+from pythermalcomfort.charts import adaptive_chart
```
Similar to below, where and how do want to add that description? As inline commments?

FedericoTartarini replied:
after the code if there is space, for example
```
v = 0.1  # air speed
```
or above the code for long comment or code string
```
# very long comment .......................
x = "owewodpeowpdowndewopnedowendwoednwoendowwepodn"
```

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+            fill="toself",
+            fillcolor="rgba(0, 128, 0, 0.2)",
+            line=dict(color="rgba(255,255,255,0)"),
+            name="90% Acceptability",
```
Yes, that is a good point. Are you thinking Enums?

FedericoTartarini replied:
yes, let's discuss that either Enum or @dataclass

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+
+    fig.add_trace(
+        go.Scatter(
+            x=np.concatenate([t_outdoor, t_outdoor[::-1]]),
```
I am not extracting values here, just reversing a list. That should be okay, right?

FedericoTartarini replied:
My bad because I didn't see this. However, I still think we need to add the comment on why we are reversing the list. It is not apparent to me. Why we're doing it.

FedericoTartarini commented:
@t-kramer thank you for the great discussion today. It is always great to chat with you. Below is code snippet of the dataclasses. Inside the data class so we can also embed functions if we want for example to return the colors of all the risk categories

```
from pydantic import BaseModel
from typing import List, Optional
from enum import Enum
from dataclasses import fields, dataclass, field
from datetime import datetime
from pytz import all_timezones

# BaseModel class for recommendations
class Recommendation(BaseModel):
    text: str  # Text of the recommendation
    icon: Optional[str] = None  # Optional icon for the recommendation


# BaseModel class for risk values
class RiskValues(BaseModel):
    risk_name: str
    color: str
    threshold: List[float]
    description: Optional[str] = None
    recommendations: Optional[List[Recommendation]] = None
    control_strategies: Optional[List[str]] = None
    value_risk: Optional[float] = None


# BaseModel class for risk models values
class RiskModelsValues(BaseModel):
    risks: List[RiskValues]  # List of risk values associated with the model
    name_model: Optional[str] = None  # Optional name of the risk model


# Enum class for different risk models
class RiskModels(Enum):
    world_rugby = RiskModelsValues(
        name_model="World Rugby",  # Name of the risk model
        risks=[
            # Defining various risk levels with their associated values and recommendations
            RiskValues(
                risk_name="Low",
                recommendations=[
                    Recommendation(
                        text="No control required", icon="carbon:checkmark-outline"
                    )
                ],
                threshold=[0, 100],
                color="#0096FF",
            ),
```

FedericoTartarini reviewed (commented):
added the comments I wrote during the meeting

t-kramer committed:
build: merge remote-tracking branch
'origin/development' into plotting-feature

t-kramer commented on examples/plot_heatmap_utci.py:
```diff
@@ -0,0 +1,78 @@
+
+
+weather_data["UTCI"] = weather_data.apply(
+    lambda row: (
```
Adjusted.

t-kramer commented on examples/plot_heatmap_utci.py:
```diff
@@ -0,0 +1,78 @@
+)
+
+
+def categorize_utci(x, si_ip):
```
Done.

t-kramer commented on examples/plot_heatmap_hi.py:
```diff
@@ -0,0 +1,63 @@
+)
+
+
+fig = heatmap(weather_data, "HI", global_local="global", si_ip="si", show_summary=True)
```
Okay, will adopt from now and adjust in other functions.

t-kramer commented on examples/plot_heatmap_hi.py:
```diff
@@ -0,0 +1,63 @@
+
+def categorize_heat_index(x, si_ip):
+
+    if si_ip.lower() == "ip":
```
I just created an issue on this. Let me know how you want to proceed. We can continue the conversation there.

t-kramer commented on examples/plot_heatmap_hi.py:
```diff
@@ -0,0 +1,63 @@
+
+weather_data = weather_file.dataframe
+
+weather_data["Hour"] = weather_data["Hour"] - 1
```
Done and will be adopted for similar lines of code.

t-kramer commented on examples/plot_heatmap_utci.py:
```diff
@@ -0,0 +1,78 @@
+
+weather_file = epw()  # https://github.com/building-energy/epw
+weather_file.read(
+    "test_data/USA_AZ_Phoenix-Sky.Harbor.Intl.AP.722780_TMYx.2009-2023.epw"
```
Done and added to `utilities.py` as `epw_import`.

t-kramer commented on examples/plot_heatmap_hi.py:
```diff
@@ -0,0 +1,63 @@
+weather_data["Month Name"] = weather_data["DateTime"].dt.strftime("%b")
+
+
+weather_data["HI"] = weather_data.apply(
```
Replaced the lambda function with a more straightforward implementation.

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+            x=t_outdoor,
+            y=t_comfort,
+            mode="lines",
+            name="Comfort Temperature",
```
Can also be an argument passed to an instance of a class.

t-kramer commented on pythermalcomfort/charts/adaptive_chart.py:
```diff
@@ -0,0 +1,225 @@
+
+    """
+
+    var = "adaptive"
```
Done.

t-kramer committed:
build: implement dataclasses for adaptive charts

Commit covers a first implementation of dataclasses for adaptive charts.
While not completely implemented, I think the idea is clear. Based on
review we can implement a similar structure for the other charts.

The following additional, minor changes were made:
- implemented return of stress_category for heat_index_rothfusz
- added >= 27 °C criterion for heat_index_rothfusz, returns NaN if not met
- adjusted plot_heatmap_hi accordingly
- slight modifications to application of comfort models to dataframes:
conversion to np.array to longer needed

t-kramer commented:
@FedericoTartarini I just pushed my latest changes and will share a video link shortly to explain them in more detail (avoiding an overly long description here).

## Summary of Changes:

1. Dataclasses for Adaptive Chart:
Partially implemented dataclasses for the adaptive chart. Most settings for the base plot and traces (data samples, overlays, and summaries) are now defined in centralized class definitions with methods.

2. Modular Plot Wrappers:
Refactored the plot functions into wrappers that host class instances, making the design more modular. Adding or removing overlays is now user-configurable and straightforward.

3. Tweaks to `heat_index_rothfusz`:
Made minor adjustments to allow the function to return `stress_categories` (similar to the `utci` model function, as discussed in #168).

Especially on the first two points, I'd like you to review before I continue implementing the some structure for the other charts. 

### Next Steps:
Most open comments will be resolved once dataclasses are fully implemented across all functions. However, a few items still need clarification/discussion/additional changes:

- Plotting Functions for Model Variants: Should we create separate plotting functions for different versions (e.g., EN vs. ASHRAE), or implement logic based on the `units` argument?
- Responsiveness of charts to `units` Argument: Do you have a preferred approach for handling this?
- Docstrings and Comments: I'll refine and expand these once we decide on the general structure to improve the clarity of the code.
- Better Input Data for Adaptive Model: I still need to make the switch to a more suitable dataset than the ASHRAE Global Thermal Comfort Database II (e.g., one allowing calculation of the running mean outdoor temperature)

Let me know your thoughts on these points and how you'd like to proceed!

FedericoTartarini commented:
Great idea recording the video and great work. Brief comments are below:
1. please create a separate pull request for the Heat Index so I can add it to pythermalcomfort. 
    1. I want to remove the units as input as much as possible especially in pythermalcomfort since are hard to manage and people can figure out themselves how to convert to C. Hence please remove the units from the heat_idex_code and just implement the mapping changes. 
    2. Please also add tests for this new functionality.
4. I really like the dataclass approach. 
    1. Not sure we need a base class since basically we need to define everything anyway when we create the AdaptiveBase class. Naming will be important and I would rename `AdaptiveBase` to `AdaptivePlot`.
    2. same comment with the Trace class, do we need a base class? can we just have Observations?
    3. Observations should be called ScatterPlot? then the function called `generate_trace` should be called plot()
    4. I do not like the code row=1 if base.show_summary is not very intuitive.
    5. How will people be able to change the classes?
    6. I am not sure why we need type within the classes
    7. we need to make it very explicity that people will have the columns names defined in line 123
    8. maybe we should not have the functions within the class
9. I will need to play a bit with the code
10. Please do not use `lambda`.
11. The adaptive function should not be included inside the plot but the users should pass these values. I did not get the point :)
12. In future please share the code.

Here is the [video](https://unisyd-my.sharepoint.com/:v:/g/personal/federico_tartarini_sydney_edu_au/ESDyaMOIx1VAtFNvf-Qx7A0ByfpY_WCR5GAnoVI32FmuLg?e=oQIgVe) I recorded.

t-kramer commented:
@FedericoTartarini Thanks for the feedback. I'll keep working on it.

Quick point: I did push the code yesterday before updating this PR, so you should be able to see the changes? Please check again.

FedericoTartarini commented:
sorry, I think I have heard in the video that you mentioned that you did not push the code so I did not check the pull request. Please let me know if my comments make sense and please let me know if you do not agree with them.

stefanoschiavon commented:
@t-kramer 

Suggestions to improve the graphs for the adaptive thermal comfort model chart:
<img width="798" height="387" alt="Screenshot 2025-10-03 at 5 17 22 PM" src="https://github.com/user-attachments/assets/63f54542-946e-4dd4-96ae-967275904955" />

1. use the same green colors as the actual tool
2. Reduce the number of x-labels, from one every 2°C, to 5°C at a minimum. 
3. Reduce the number of y-labels, from one every 2°C, to 5°C at a minimum. 
4. Remove the "comfort temperature", it is not a concept valid within 55
5. VIP. Acceptable for 80% must include also all the people that is accpetable for 90%, so it should be 64.6%. 
6. Consider rounding at zero decimal points. 
7. Change from "Out for range" to " Our of applicability range"
8. In general, reduce word capitalization as much as possible. 

Suggestions to improve the graphs for the PMV psychrometric model chart:
<img width="817" height="429" alt="Screenshot 2025-10-03 at 5 24 11 PM" src="https://github.com/user-attachments/assets/ca1a7a11-8cf6-4583-87aa-3dfb4417cf56" />

1. Remove the neutral PMV, this concept does not exist in ASHRAE. 
2. Change from "Too Hot" to "Warmer than comfortable" or "Uncomfortably warm". I prefer the former. Same for "Too cold"
3. Change from "Out for range" to " Our of applicability range"

Suggestions to improve the graphs for the HI:
<img width="815" height="379" alt="Screenshot 2025-10-03 at 5 27 22 PM" src="https://github.com/user-attachments/assets/863d5672-5a8f-424d-94d4-fe4aa7a5f444" />

1. VIP. Remove "No Risk", this category does not exist in US. 
2. You could use the colors from the NYT https://www.nytimes.com/interactive/2022/us/heat-wave-map-tracker.html or the one from NOAA: https://www.noaa.gov/media/image_download/79db348b-7b5a-4e19-94e0-c0a87209a8ba. I tend to trust NYT more and it seems that your colors are closer to NOAA 

For the UTCI and HI index show the hours every 3 o 6 hours, not every 2.

FedericoTartarini commented:
Thank you @stefanoschiavon for the comments. We will incorporate them in the new plotting functions. However, I was hoping to get your feedback on this discussion https://github.com/CenterForTheBuiltEnvironment/pythermalcomfort/discussions/252 not on this pull request.
